src/upgrade.py

```diff
@@ -0,0 +1,219 @@
+            )
+
+        try:
+            self._pre_upgrade_prepare()
+        except MySQLSetClusterPrimaryError:
+            raise ClusterNotReadyError(
+                message=fail_message,
+                cause="Failed to set primary",
+                resolution="Check the cluster status",
+            )
+        except KubernetesClientError:
+            raise ClusterNotReadyError(
+                message=fail_message,
+                cause="Failed to patch statefulset",
+                resolution="Check kubernetes access policy",
+            )
+
+    @override
+    def log_rollback_instructions(self) -> None:
+        """Log rollback instructions."""
+        logger.info(
+            "Run `juju refresh --revision <previous-revision> mysql-k8s` to initiate the rollback"
+        )
+        logger.info("and `juju run-action mysql-k8s/leader resume-upgrade` to resume the rollback")
+
+    def _pre_upgrade_prepare(self) -> None:
+        """Pre upgrade routine for MySQL.
+
+        Set primary to the first unit to avoid switchover during upgrade and
+        patch statefulset `spec.updateStrategy.rollingUpdate.partition` to the last unit.
+        """
+        if self.charm._mysql.get_primary_label() != f"{self.charm.app.name}-0":
+            # set the primary to the first unit for switchover mitigation
+            self.charm._mysql.set_cluster_primary(
+                self.charm._get_unit_fqdn(f"{self.charm.app.name}/0")
+            )
+
+        self.charm.k8s_helpers.set_rolling_update_partition(partition=self.highest_ordinal)
+
+    def _on_resume_upgrade(self, event: ActionEvent) -> None:
+        """Handle resume upgrade action.
+
+        Continue the upgrade by setting the partition to the next unit.
+        """
+        if not self.charm.unit.is_leader():
+            event.fail("Action only valid on leader unit")
+            return
+        fail_message = "Nothing to resume, upgrade stack unset"
+        if self.upgrade_stack:
+            try:
+                next_partition = self.upgrade_stack[-1]
+                event.set_results({"message": f"Upgrade will resume on unit {next_partition}"})
+                self.charm.k8s_helpers.set_rolling_update_partition(partition=next_partition)
+                return
+            except IndexError:
+                fail_message = "Nothing to resume, empty upgrade stack"
+            except KubernetesClientError:
+                fail_message = "Cannot set rolling update partition"
+        event.fail(fail_message)
+
+    def _on_pebble_ready(self, event) -> None:
+        """Handle pebble ready event.
+
+        Confirm that unit is healthy and set unit completed.
+        """
+        if not self.peer_relation:
+            logger.debug("Peer relation not yet ready on unit. Deferring event")
+            event.defer()
+            return
+
+        if self.peer_relation.data[self.charm.unit].get("state") in ["ready", "upgrading"]:
```

Comment:
On k8s, units should be on `upgrading` state. We skipped the `ready` state for k8s on https://github.com/canonical/data-platform-libs/pull/77 and I also saw the same in you `_on_upgrade_charm ` method.